docs/proposals/improve-status-in-CAPI-resources.md

```diff
@@ -0,0 +1,958 @@
+| `UpdatedReplicas`                    | `UpToDateReplicas` (renamed)                                | `UpToDateReplicas`                  |
+| `ExprimentalReadyReplicas` (new)     | `ReadyReplicas` (renamed)                                   | `ReadyReplicas`                     |
+| `ExprimentalAvailableReplicas` (new) | `AvailableReplicas` (renamed)                               | `AvailableReplicas`                 |
+|                                      | `BackCompatibilty` (new)                                    | (removed)                           |
+| `ReadyReplicas` (deprecated)         | `BackCompatibilty.ReadyReplicas` (renamed) (deprecated)     | (removed)                           |
+| `AvailableReplicas` (deprecated)     | `BackCompatibilty.AvailableReplicas` (renamed) (deprecated) | (removed)                           |
+| `Phase` (deprecated)                 | `BackCompatibilty.Phase` (renamed) (deprecated)             | (removed)                           |
+| `FailureReason` (deprecated)         | `BackCompatibilty.FailureReason` (renamed) (deprecated)     | (removed)                           |
+| `FailureMessage` (deprecated)        | `BackCompatibilty.FailureMessage` (renamed) (deprecated)    | (removed)                           |
+| `Conditions`                         | `BackCompatibilty.Conditions` (renamed) (deprecated)        | (removed)                           |
+| `ExperimentalConditions` (new)       | `Conditions` (renamed)                                      | `Conditions`                        |
+| other fields...                      | other fields...                                             | other fields...                     |
+
+Notes:
+- The `BackCompatibilty` struct is going to exist in v1beta2 types only until v1beta1 removal (9 months or 3 minor releases after v1beta2 is released/v1beta1 is deprecated, whichever is longer).
+  Fields in this struct are used for supporting down conversions, thus providing users relying on v1beta1 APIs additional buffer time to pick up the new changes.
+
+#### MachineDeployment (New)Conditions
+
+| Condition        | Note                                                                                                                                                                                                                                                   |
+|------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `Available`      | True if the MachineDeployment has minimum availability according to parameters specified in the deployment strategy, e.g. If using RollingUpgrade strategy, availableReplicas must be greater or equal than desired replicas - MaxUnavailable replicas |  
```

Comment:
> availableReplicas must be greater or equal than desired replicas - MaxUnavailable replicas

Is the `RollingUpgrade` strategy the right source of this information? I may only want one node to be intentionally unavailable at a time during a planned rollout, but expect to be able to tolerate two nodes being unavailable (e.g., in case some failure occurs during an upgrade). If I've planned for that, maybe I wouldn't want to see `Available: false` if I've got two nodes unavailable?

I also wonder about possible interactions here with scaling operations, including those triggered by Cluster Autoscaler: if the desired number of replicas increases a bunch, seeing `ScalingUp: true` and `Available: false` seems like it could be unnecessarily alarming. From a glance, it seems like something bad happened during the scale up, rather than an expected consequence of a scale up.

Reply:
Sharing a few initial thoughts.

- MD RollingUpgrade strategy IMO is the current source of this information, because this is the single point of control for this group of machines and everything below should be considered ephemeral
- The idea of unbalanced/more sophisticated RollingUpgrade strategies could be something interesting to think about, even if the downside of it is that it adds complexity to a behaviour that is nicely consistent with K8s Deployments.
- `ScalingUp: true` and `Available: false` does not seem alarming to me _if_ this is considered together with current and desired replicas, because the last two info should give you an hint about the fact that someone or something asked for additional capacity (and those info are all available in the same place)

However, frankly speaking, even if interesting, discussing how MD performs rolling upgrades + if and how we should improve this seems a little bit out of scope of this proposal (also MD.status.Conditions[Available] one of the conditions we are not changing in this proposal 😅 ).

But If we want to continue this discussion happy to do it on a separated issue

Reply:
After thinking a bit about this.

I could also see the case for deciding available just based on if "available replicas" >= "desired replicas".

I think this is already the case if someone uses MaxUnavailable = 0, or if folks would use a different strategy.

Also really not sure if the updateStrategy should be the source of truth for the decision what "available" means on a MachineDeployment, it feels a bit out of scope for the updateStrategy.

Reply:
But agree, the current definition in this proposal is our current implementation. Maybe worth a follow-up on the umbrella issue though

Reply:
Resolving, eventually we can follow up on ideas for improvements on rolling upgrades on a separated issue